trie/trie.py

```diff
@@ -378,3 +397,250 @@ def __delitem__(self, key):
 
     def __contains__(self, key):
         return self.exists(key)
+
+
+class BinaryTrie():
+    def __init__(self, db, root_hash=BLANK_HASH):
+        self.db = db
+        validate_is_bytes(root_hash)
+        self.root_hash = root_hash
+
+    def get(self, key):
+        """
+        Fetches the value with a given keypath from the given node.
+
+        Keyp will be encoded into binary array format first
+        """
+        validate_is_bytes(key)
+
+        return self._get(self.root_hash, encode_to_bin(key))
+
+    def _get(self, node_hash, keypath):
+        """
+        Note: keypath should be in binary array format, i.e., encoded by encode_to_bin()
+        """
+        # Empty trie
+        if node_hash == BLANK_HASH:
+            return None
+        nodetype, left_child, right_child = parse_node(self.db[node_hash])
+        # Key-value node descend
+        if nodetype == LEAF_TYPE:
+            return right_child
+        elif nodetype == KV_TYPE:
+            # Keypath too short
+            if not keypath:
+                return None
+            if keypath[:len(left_child)] == left_child:
+                return self._get(right_child, keypath[len(left_child):])
+            else:
+                return None
+        # Branch node descend
+        elif nodetype == BRANCH_TYPE:
+            # Keypath too short
+            if not keypath:
+                return None
+            if keypath[:1] == BYTE_0:
+                return self._get(left_child, keypath[1:])
+            else:
+                return self._get(right_child, keypath[1:])
+
+    def set(self, key, value):
+        """
+        Sets the value at the given keypath from the given node
+
+        Keyp will be encoded into binary array format first
+        """
+        validate_is_bytes(key)
+        validate_is_bytes(value)
+
+        self.root_hash = self._set(self.root_hash, encode_to_bin(key), value)
+
+    def _set(self, node_hash, keypath, value):
```

Comment:
This function body is about 150 lines long.

Can we split it up a bit.  Maybe split the body up into individual functions for each node type to reduce the size a bit?

```python
if nodetype == LEAF_TYPE:
    self._set_leaf_node(...)
elif nodetype == KV_TYPE:
    self._set_kv_node(...)
elif nodetype == BRANCH_TYPE:
    self._set_branch_node(...)
```